File: app/src/main/java/org/jellyfin/androidtv/ui/search/LeanbackSearchFragment.java
package org.jellyfin.androidtv.ui.search;

import static org.koin.java.KoinJavaComponent.inject;

import android.os.Bundle;

import androidx.leanback.app.SearchSupportFragment;
import androidx.leanback.widget.ListRow;

import org.jellyfin.androidtv.data.service.BackgroundService;
import org.jellyfin.androidtv.ui.browsing.CompositeSelectedListener;
import org.jellyfin.androidtv.ui.itemhandling.BaseRowItem;
import org.jellyfin.androidtv.ui.itemhandling.ItemLauncher;
import org.jellyfin.androidtv.ui.itemhandling.ItemRowAdapter;

import kotlin.Lazy;

public class LeanbackSearchFragment extends SearchSupportFragment {
    protected CompositeSelectedListener mSelectedListener = new CompositeSelectedListener();

    private Lazy<BackgroundService> backgroundService = inject(BackgroundService.class);

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        backgroundService.getValue().attach(requireActivity());

        // Create provider
        SearchProvider searchProvider = new SearchProvider(getContext(), getActivity().getIntent().getBooleanExtra("MusicOnly", false));
        setSearchResultProvider(searchProvider);

        // Create click listener
        setOnItemViewClickedListener((itemViewHolder, item, rowViewHolder, row) -> {
            if (!(item instanceof BaseRowItem)) return;

            ItemLauncher.launch((BaseRowItem) item, (ItemRowAdapter) ((ListRow) row).getAdapter(), ((BaseRowItem) item).getIndex(), getActivity());
        });

        setOnItemViewSelectedListener(mSelectedListener);
        mSelectedListener.registerListener((itemViewHolder, item, rowViewHolder, row) -> {
            if (!(item instanceof BaseRowItem)) {
                backgroundService.getValue().clearBackgrounds();
            } else {
                BaseRowItem rowItem = (BaseRowItem) item;
                backgroundService.getValue().setBackground(rowItem.getSearchHint());
            }
        });
    }
}

File: app/src/main/java/org/jellyfin/androidtv/ui/search/LeanbackSearchFragment.kt
package org.jellyfin.androidtv.ui.search

import android.os.Bundle
import android.view.View
import androidx.leanback.app.SearchSupportFragment
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import org.koin.android.ext.android.inject
import org.koin.androidx.viewmodel.ext.android.viewModel
import org.koin.core.parameter.parametersOf

class LeanbackSearchFragment : SearchSupportFragment(), SearchSupportFragment.SearchResultProvider {
	private val viewModel: SearchViewModel by viewModel()

	private val searchFragmentDelegate: SearchFragmentDelegate by inject {
		parametersOf(requireContext())
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		setSearchResultProvider(this)
		setOnItemViewClickedListener(searchFragmentDelegate.onItemViewClickedListener)
		setOnItemViewSelectedListener(searchFragmentDelegate.onItemViewSelectedListener)
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		super.onViewCreated(view, savedInstanceState)

		viewModel.searchResultsFlow
			.onEach { searchFragmentDelegate.showResults(it) }
			.launchIn(lifecycleScope)

		val query = arguments?.getString(SearchFragment.EXTRA_QUERY)
		if (!query.isNullOrBlank()) setSearchQuery(query, true)
	}

	override fun getResultsAdapter() = searchFragmentDelegate.rowsAdapter
	override fun onQueryTextChange(query: String): Boolean = viewModel.searchDebounced(query)
	override fun onQueryTextSubmit(query: String): Boolean = viewModel.searchImmediately(query)

}
